[figure_1](figure_1), [figure_2](figure_2), [figure_2_exp](figure_2_exp), [figure_aligners](figure_aligners), [figure_aligners_2](figure_aligners_2), [figure_merged](figure_merged), [figure_mm10](figure_mm10), [figure_parameters](figure_parameters): `BED` files of exclusion lists used in [figures.R](../figures.R).

[cds](cds), [genes](genes), [exons](exons): `BED` files of various biological features used in [figures.R](../figures.R).

[hg38_gaps](hg38_gaps), [mm10_gaps](mm10_gaps): `BED` files of gaps used in [figures.R](../figures.R).

[summary_36](summary_36), [summary_101](summary_101): `TSV` files generated with deepTools [`multiBamSummary`](https://gensoft.pasteur.fr/docs/deepTools/3.4.1/content/tools/multiBamSummary.html) for our 36bp ChIP-seq control data and 101bp ChIP-seq control data.

[tf_correlations](tf_correlations): `TSV` files of transcription factor (GM12878) pearson correlation matrices using various exclusion lists. Generated with deepTools [plotCorrelation](https://gensoft.pasteur.fr/docs/deepTools/3.4.1/content/tools/plotCorrelation.html).

[WGS](WGS): `CSV` files of SNPs and INDELs gained or lost when using hg38 with the extra sponge regions.

[bam_accession_table.csv](bam_accession_table.csv): Metadata on the the ENCODE ChIP-seq control BAMs used to generate exclusion sets.

[bulkRNASeq.zip](bulkRNASeq.zip): A zipped folder of bulk RNASeq data used in [figures.R](../figures.R).

[donor_accession_table.csv](bam_accession_table.csv): Metadata on the the ENCODE ChIP-seq control donors used to generate exclusion sets.

[download_tf_data.sh](download_tf_data.sh): A script to download transcription factor data from ENCODE.

[metadata.tsv](metadata.tsv): Other metadata on the the ENCODE ChIP-seq control data used to generate exclusion sets.

[tf_urls.txt](tf_urls.txt): URLs from ENCODE for transcription factor data used in our analysis.

[gencode.v46.annotation.gff3.gz](gencode.v46.annotation.gff3.gz): GRCh38.p14 gene annotations from gencode found [here](https://www.gencodegenes.org/human/release_46).
